js_modules/dagster-ui/packages/ui-core/src/asset-data/AssetLiveDataThreadManager.tsx

```diff
@@ -0,0 +1,241 @@
+        assetsToFetch.push({path: key.split('/')});
+      } else {
+        assetsWithoutData.push({path: key.split('/')});
+      }
+    }
+
+    // Prioritize fetching assets for which there is no data in the cache
+    return assetsWithoutData.concat(assetsToFetch).slice(0, BATCH_SIZE);
+  }
+
+  public areKeysRefreshing(assetKeys: AssetKeyInput[]) {
+    for (const key of assetKeys) {
+      const stringKey = tokenForAssetKey(key);
+      if (!this.lastFetchedOrRequested[stringKey]?.fetched) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  private getAllObservedKeys() {
+    const threads = Object.values(this.threads);
+    return Array.from(new Set(threads.flatMap((thread) => thread.getObservedKeys())));
+  }
+
+  public getOldestDataTimestamp() {
+    const allAssetKeys = Object.keys(this.listeners).filter((key) => this.listeners[key]?.length);
+    let isRefreshing = allAssetKeys.length ? true : false;
+    let oldestDataTimestamp = Infinity;
+    for (const key of allAssetKeys) {
+      if (this.lastFetchedOrRequested[key]?.fetched) {
+        isRefreshing = false;
+      }
+      oldestDataTimestamp = Math.min(
+        oldestDataTimestamp,
+        this.lastFetchedOrRequested[key]?.fetched ?? Infinity,
+      );
+    }
+    return {
+      isRefreshing,
+      oldestDataTimestamp: oldestDataTimestamp === Infinity ? 0 : oldestDataTimestamp,
+    };
+  }
+
+  public _updateCache(data: Record<string, LiveDataForNode>) {
+    Object.assign(this.cache, data);
+  }
+
+  public onDocumentVisiblityChange(isDocumentVisible: boolean) {
+    if (isDocumentVisible) {
+      if (this.isPaused) {
+        this.unpause();
+      }
+    } else if (!this.isPaused) {
+      this.pause();
+    }
+  }
+
+  private pause() {
+    this.isPaused = true;
+    Object.values(this.threads).forEach((thread) => {
+      thread.stopFetchLoop();
+    });
+  }
+
+  private unpause() {
+    this.isPaused = false;
+    Object.values(this.threads).forEach((thread) => {
+      thread.startFetchLoop();
+    });
+  }
+
+  public getCacheEntry(key: AssetKeyInput) {
+    return this.cache[tokenForAssetKey(key)];
+  }
+
+  public _markAssetsRequested(assetKeys: AssetKeyInput[]) {
+    const requestTime = Date.now();
+    assetKeys.forEach((key) => {
+      this.lastFetchedOrRequested[tokenForAssetKey(key)] = {
+        requested: requestTime,
+      };
+    });
+    this.onUpdatedOrUpdating();
+  }
+
+  public _unmarkAssetsRequested(assetKeys: AssetKeyInput[]) {
+    assetKeys.forEach((key) => {
+      delete this.lastFetchedOrRequested[tokenForAssetKey(key)];
+    });
+  }
+
+  public _updateFetchedAssets(assetKeys: AssetKeyInput[], data: Record<string, LiveDataForNode>) {
+    const fetchedTime = Date.now();
+    assetKeys.forEach((key) => {
+      this.lastFetchedOrRequested[tokenForAssetKey(key)] = {
+        fetched: fetchedTime,
+      };
+    });
+    Object.entries(data).forEach(([key, assetData]) => {
```

Comment:
Not a big deal at all but this `key` is a tokenForAssetKey, and the previous loop's `key` is a real AssetKey. Would be kinda nice to keep the terminology explicit somehow if we could.

 I've thought about trying to do something wild and define `AssetKey.toString` as `tokenForAssetKey` but that's just too gross 😭

Reply:
Yeah that's a good point, I can actually combine the two loops